sfrCore/model/instance.py

```diff
@@ -323,16 +323,31 @@ def updateAgent(self, agent):
             logger.warning('Unable to read agent {}'.format(agent['name']))
 
     def addIdentifiers(self):
-        for iden in self.tmp_identifiers:
-            if iden['type'] == 'isbn':
-                self.fetchUnglueitSummary(iden['identifier'])
-            self.upsertIdentifier(iden)
+        """This method takes a list of potential new identifiers to associate
+        with this instance. New identifiers are added and if they are ISBNs
+        a summary from unglue.it is checked for.
+
+        Existing identifiers are skipped.
+        """
+        identifiers = {
+            self.upsertIdentifier(iden) for iden in self.tmp_identifiers
+        }
+
+        # This removes all existing identifiers from set, removing unnecessary
+        # operations
+        identifiers.difference_update(self.identifiers)
+
+        # Add new identifiers to set and check for summaries associated with
+        # new identifiers being associated with this instance
+        for iden in list(identifiers):
+            self.identifiers.add(iden)
+            if iden.type == 'isbn':
+                self.fetchUnglueitSummary(iden.isbn[0].value)
 
     def upsertIdentifier(self, iden):
         try:
-            self.identifiers.add(
-                Identifier.returnOrInsert(self.session, iden)
-            )
+            logger.debug('Adding {}'.format(iden['identifier']))
```

Comment:
Not super familiar with Python but is this log statement being run for all the identifiers?

Reply:
Yes, for now. But the logger is configurable with an environment variable so it will be (or can be) silenced in production.

Its helpful in development because it helps identify if an issue is with a single identifier or is a bug in the method

Reply:
sounds good!